compiler/rustc_lint/messages.ftl

```diff
@@ -440,17 +440,19 @@ lint_non_local_definitions_cargo_update = the {$macro_kind} `{$macro_name}` may
 
 lint_non_local_definitions_deprecation = this lint may become deny-by-default in the edition 2024 and higher, see the tracking issue <https://github.com/rust-lang/rust/issues/120363>
 
-lint_non_local_definitions_impl = non-local `impl` definition, they should be avoided as they go against expectation
+lint_non_local_definitions_impl = non-local `impl` definition, `impl` blocks should be written at the same level as their item
     .help =
         move this `impl` block outside the of the current {$body_kind_descr} {$depth ->
             [one] `{$body_name}`
            *[other] `{$body_name}` and up {$depth} bodies
         }
-    .non_local = an `impl` definition is non-local if it is nested inside an item and may impact type checking outside of that item. This can be the case if neither the trait or the self type are at the same nesting level as the `impl`
-    .exception = one exception to the rule are anon-const (`const _: () = {"{"} ... {"}"}`) at top-level module and anon-const at the same nesting as the trait or type
+    .without_trait = methods and assoc const are still usable outside the current expression, only `impl Local` and `impl dyn Local` are local and only if the `Local` type is at the same nesting as the `impl` block
+    .with_trait = an `impl` definition is non-local if it is nested inside an item and may impact type checking outside of that item, which can be the case if neither the trait or the self type are at the same nesting level as the `impl`
```

Comment:
hmm so I think we could start more clearly like this:
```
    .with_trait = an `impl` is always non-local
```
with more to it, but "if" is not the next word.

my reasoning is: the impl is non-local by definition, right? it "always is", even if nested. *however*, this only *matters* (affects code in a visible way) if at least one of (list of other considerations we've already discussed to death) applies. so we're choosing to complain about the non-locality in certain cases (roughly "when we believe it matters").

Reply:
we might want to try a "yes, *and*" approach like:
```suggestion
    .with_trait = an `impl` is always non-local and may allow access by type inference outside this nesting item if a type and trait is visible
```
am I saying that right? "nesting item" is what I'm saying to try to describe "the thing that the nest*ed* item is *nesting in*". I'm hoping it is somewhat clear from context in actual experience.

Reply:
I like the idea of starting with "is always non-local", but the suggested text implies that the user is trying to "hide" the type, but the note was trying to be more general to also cover cases where neither are being hidden,
```rust
trait Trait {}
struct Type;

fn main() {
    impl Trait for type {}
}
```

what about?
```suggestion
    .with_trait = an `impl` is always non-local, and even when nested inside an item it may impact type checking outside of that item, which can be the case if neither the trait or the self type are at the same nesting level as the `impl`
```

Reply:
sure!

Reply:
Done in https://github.com/rust-lang/rust/pull/125089/commits/b4024cccaf0ab5a0dbdc88753b618f652bcbda7e.